File: Sliding_Window/121_best_time_to_buy_and_sell_stock/121.py


def maxProfit(prices):        
        
        min_price = prices[0]
        max_profits = 0
        
        for i in range(1, len(prices)):
            
            min_price = min(min_price, prices[i])
            max_profits = max(prices[i] - min_price, max_profits)
            
        return max_profits


if __name__ == "__main__":
    prices = [7,1,5,3,6,4]
    print("result : ", maxProfit(prices))

File: Sliding_Window/121_best_time_to_buy_and_sell_stock/121.java
/**
 * Brute Force Approach
 * T: O(N ^ 2)
 * S: O(1)
 *
 * One Pass
 * T: O(N)
 * S: O(1)
 */
class Solution {
    public int maxProfit(int[] prices) {

        int minPrice = Integer.MAX_VALUE;
        int maxValue = 0;

        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            }

            if ((price - minPrice) > maxValue) {
                maxValue = price - minPrice;
            }
        }

        return maxValue;
    }
}